cpp/negentropy_wrapper.c

```diff
@@ -298,3 +282,153 @@ void free_result(result* r){
         free((void *)r->error);
     }
 }
+
+/*SubRange specific functions
+TODO: These and above methods need to be optimized to reduce code duplication*/
+
+void* subrange_new(void* storage, uint64_t startTimeStamp, uint64_t endTimeStamp){
+    negentropy::storage::BTreeMem* st = reinterpret_cast<negentropy::storage::BTreeMem*>(storage);
+    negentropy::storage::SubRange* subRange = NULL;
+    try {
+    subRange = new negentropy::storage::SubRange(*st, negentropy::Bound(startTimeStamp), negentropy::Bound(endTimeStamp));
+    } catch (negentropy::err e){
+        return NULL;
+    }
+    return subRange;
+}
+
+void subrange_delete(void* range){
+    negentropy::storage::SubRange* subRange = reinterpret_cast<negentropy::storage::SubRange*>(range);
+    delete subRange;
+}
+
+int subrange_size(void* range){
+    negentropy::storage::SubRange* subrange = reinterpret_cast<negentropy::storage::SubRange*>(range);
+    return subrange->size();
+}
+
+void negentropy_subrange_delete(void* negentropy){
+    Negentropy<negentropy::storage::SubRange>* ngn_inst = reinterpret_cast<Negentropy<negentropy::storage::SubRange>*>(negentropy);
+    delete ngn_inst;
+}
+
+void* negentropy_subrange_new(void* subrange, uint64_t frameSizeLimit){
+    //TODO: Make these typecasts into macros??
+    negentropy::storage::SubRange* sub_range;
+    //TODO: reinterpret cast is risky, need to use more safe type conversion.
+    sub_range = reinterpret_cast<negentropy::storage::SubRange*>(subrange);
+
+    Negentropy<negentropy::storage::SubRange>* ne;
+    try{
+        ne = new Negentropy<negentropy::storage::SubRange>(*sub_range, frameSizeLimit);
+    }catch(negentropy::err e){
+        return NULL;
+    }
+    return ne;
+}
+
+// Returns -1 if already initiated.
+int negentropy_subrange_initiate(void* negentropy, result* result){
+    Negentropy<negentropy::storage::SubRange>* ngn_inst;
+    ngn_inst = reinterpret_cast<Negentropy<negentropy::storage::SubRange>*>(negentropy);
+
+    std::string* output = new std::string();
+    try {
+        *output = ngn_inst->initiate();
+/*         std::cout << "output of initiate is, len:" << output->size() << ", output:";
+        printHexString(std::string_view(*output)); */
+    } catch(negentropy::err e){
+        //std::cout << "Exception raised in initiate " << e.what() << std::endl;
```

Comment:
It is interesting to capture the error and give feedback to the caller

Reply:
will do error handling in separate PR, added a TODO for now.